<a href="https://colab.research.google.com/github/nakulreddy0107/NakulReddy_INFO5731_Spring2025/blob/main/INFO5731_Assignment_1_Spring_2025_(4).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment - 1**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.


1.  Question 1-5 using AI assistance is strictly prohibted
2.  Question 6-7 AI assitance is recommended by following guidelines.

Guidelines and expectations file can be found in week 2 module.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [ ]:
# Your code here

import math

print("r\tV")
for r in range(2, 21, 2):
    V = (2/3) * math.pi * (r ** 3)
    print(f"{r}\t{V:.2f}")

r	V
2	16.76
4	134.04
6	452.39
8	1072.33
10	2094.40
12	3619.11
14	5747.02
16	8578.64
18	12214.51
20	16755.16


# Question 2 (10 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [ ]:
import math
import random
numbers = [random.randint(50, 150) for _ in range(12)]
ascending_n = sorted(numbers)
descending_n = sorted(numbers, reverse=True)
overall_average = sum(numbers) / len(numbers)
asc_first6_avg = sum(ascending_n[:6]) / 6
desc_first6_avg = sum(descending_n[:6]) / 6

print("12 Random Numbers:", numbers)
print("Ascending Order of Numbers:", ascending_n)
print("Descending Order of Numbers:", descending_n)
print("Overall Average:", round(overall_average, 2))
print("Average of First 6 (Ascending Order):", round(asc_first6_avg, 2))
print("Average of First 6 (Descending Order):", round(desc_first6_avg, 2))




12 Random Numbers: [137, 133, 135, 67, 61, 52, 74, 118, 99, 89, 106, 92]
Ascending Order of Numbers: [52, 61, 67, 74, 89, 92, 99, 106, 118, 133, 135, 137]
Descending Order of Numbers: [137, 135, 133, 118, 106, 99, 92, 89, 74, 67, 61, 52]
Overall Average: 96.92
Average of First 6 (Ascending Order): 72.5
Average of First 6 (Descending Order): 121.33


# Question 3 (10 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [ ]:
# Your code here
import string
from collections import Counter

def get_sentences():
    sentence1 = input("Enter the first sentence: ")
    sentence2 = input("Enter the second sentence: ")
    return sentence1, sentence2

def sentence_to_list(sentence):
    return sentence.split()

def find_common_words(list1, list2):
    return list(set(list1) & set(list2))

def remove_punctuation(sentence):
    return ''.join(ch for ch in sentence if ch not in string.punctuation)

def count_vowels(sentence):
    vowels = "aeiouAEIOU"
    sentence_vowels = Counter(ch.lower() for ch in sentence if ch.lower() in vowels)
    return {v: sentence_vowels.get(v, 0) for v in "aeiou"}

def main():
    sentence1, sentence2 = get_sentences()

    word_list1 = sentence_to_list(sentence1)
    word_list2 = sentence_to_list(sentence2)

    combined_list = word_list1 + word_list2
    print("Word List:", combined_list)

    common_words = find_common_words(word_list1, word_list2)
    print("Common Words:", common_words)

    clean_sentence1 = remove_punctuation(sentence1)
    clean_sentence2 = remove_punctuation(sentence2)
    print("Sanitized Sentence 1:", clean_sentence1)
    print("Sanitized Sentence 2:", clean_sentence2)

    vowel_count1 = count_vowels(clean_sentence1)
    vowel_count2 = count_vowels(clean_sentence2)

    print("Vowel Count in Sentence 1:", ", ".join(f"{k}:{v}" for k, v in vowel_count1.items()))
    print("Vowel Count in Sentence 2:", ", ".join(f"{k}:{v}" for k, v in vowel_count2.items()))

if __name__ == "__main__":
    main()



Enter the first sentence: hi im nakul
Enter the second sentence: nakul here 
Word List: ['hi', 'im', 'nakul', 'nakul', 'here']
Common Words: ['nakul']
Sanitized Sentence 1: hi im nakul
Sanitized Sentence 2: nakul here 
Vowel Count in Sentence 1: a:1, e:0, i:2, o:0, u:1
Vowel Count in Sentence 2: a:1, e:2, i:0, o:0, u:1


# Question 4 (10 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [ ]:
import math

def  mathematical_expression(expression):
    try:
        result = eval(expression)
        return result
    except ZeroDivisionError:
        return "Error: Division by zero is not allowed."
    except Exception as e:
        return f"Error: {str(e)}"

def calculator():
    print("Supported operations: +, -, *, /, (), ** (exponentiation), sqrt (square root), % (modulus)")
    print("Type 'exit' to quit.")

    while True:
        expression = input("Enter a mathematical expression: ")
        if expression.lower() == 'exit':
            break
        expression = expression.replace('sqrt', 'math.sqrt')
        result = mathematical_expression(expression)
        print(f"Result: {result}")
calculator()


Supported operations: +, -, *, /, (), ** (exponentiation), sqrt (square root), % (modulus)
Type 'exit' to quit.
Enter a mathematical expression: quit
Result: <IPython.core.autocall.ZMQExitAutocall object at 0x78e3cc334d10>
Enter a mathematical expression: exit


# Question 5 (20 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (5 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (5 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [ ]:
 # Your code here

import pandas as pd
import scipy.stats as stats
#a
print("Question A")
df=pd.read_csv("/Assignment1_denton_housing.csv")
df[df["title_field"]=="Occupied Housing Units"]
df_1=df[df["title_field"]=="Occupied Housing Units"].copy()
df_1["consecutive year difference"]=df_1.sort_values("year")["value"].diff(-1)
print(df_1)
#b
print("Question B")
df_2=df.pivot(index="year", columns="title_field", values="value").reset_index()
df_2["VC"]= ((df_2["Vacant Housing Units"]/df_2["Total Housing Units"])*100).round(2)
df_2["OP"]= ((df_2["Occupied Housing Units"]/df_2["Total Housing Units"])*100).round(2)

for i, r in df_2.iterrows():
    print(r["year"]," ",r["VC"], "%"," ",r["OP"],"%")

#c
print("Question C")
Vacant_Housing_Units= df_2.loc[df_2['Vacant Housing Units'].idxmax()]
print("Vacant Housing Units", Vacant_Housing_Units["year"], Vacant_Housing_Units["Vacant Housing Units"] )
Occupied_Housing_Units =df_2.loc[df_2['Occupied Housing Units'].idxmax()]
print("Occupied Housing Units", Occupied_Housing_Units["year"], Occupied_Housing_Units["Occupied Housing Units"] )

#d
print("Question D")
print(stats.hmean(df_2['Total Housing Units']))



Question A
    year             title_field  value  consecutive year difference
2   2014  Occupied Housing Units  46295                          NaN
5   2013  Occupied Housing Units  43673                      -2622.0
8   2012  Occupied Housing Units  42108                      -1565.0
11  2011  Occupied Housing Units  40704                      -1404.0
14  2010  Occupied Housing Units  41007                        303.0
17  2009  Occupied Housing Units  35916                      -5091.0
20  2008  Occupied Housing Units  36711                        795.0
Question B
2008.0   10.88 %   89.12 %
2009.0   9.07 %   90.93 %
2010.0   8.14 %   91.86 %
2011.0   9.38 %   90.62 %
2012.0   6.68 %   93.32 %
2013.0   8.8 %   91.2 %
2014.0   5.73 %   94.27 %
Question C
Vacant Housing Units 2008.0 4483.0
Occupied Housing Units 2014.0 46295.0
Question D
44398.28543969185


# **For question 6 and 7 you can take AI assitance to solve.**

Make sure you follow the instructions given in class to answer these questions.


**Important points:**

1.   You must come up with at least 10 edge cases.
2.   You should provide prompts and revised prompts given to ChatGPT.
2.   Debug your code and write comments for each block of code
3.   Analyzing error will not consider as edge case.

Edge cases are like real life scienarios which are realated to the given question.

# Question 6 (20 points).
Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

In [ ]:
# Your code hereimport csv
import os
import csv
import random
import string

# Constants
FILENAME = 'inventory.csv'
FIELDS = ['ID', 'Name', 'Category', 'Price', 'Quantity']


def initialize_inventory():
    # Check if the inventory file exists, if not, create it with headers
    if not os.path.exists(FILENAME):
        with open(FILENAME, mode='w', newline='') as file:
            writer = csv.DictWriter(file, fieldnames=FIELDS)
            writer.writeheader()


def load_inventory():
    inventory = []
    with open(FILENAME, mode='r') as file:
        reader = csv.DictReader(file)
        for row in reader:
            try:
                row['Price'] = float(row.get('Price', 0))
                row['Quantity'] = int(row.get('Quantity', 0))
                # Ensure ID is unique and convert data types
                row['ID'] = int(row.get('ID', 0))
                inventory.append(row)
            except ValueError:
                print(f"Error loading product with ID {row.get('ID')}: Incorrect data format.")
    return inventory


def save_inventory(inventory):
    with open(FILENAME, mode='w', newline='') as file:
        writer = csv.DictWriter(file, fieldnames=FIELDS)
        writer.writeheader()
        writer.writerows(inventory)


def add_product(product):
    inventory = load_inventory()
    if any(item['ID'] == product['ID'] for item in inventory):
        print("Product ID already exists. Please use a unique ID.")
        return
    inventory.append(product)
    save_inventory(inventory)
    print(f"Product {product['Name']} added successfully.")


def delete_product(product_id):
    inventory = load_inventory()
    filtered_inventory = [product for product in inventory if product['ID'] != product_id]
    if len(inventory) == len(filtered_inventory):
        print("Product ID not found.")
    else:
        save_inventory(filtered_inventory)
        print(f"Product with ID {product_id} deleted successfully.")


def update_product(product_id, updated_info):
    inventory = load_inventory()
    for product in inventory:
        if product['ID'] == product_id:
            product.update(updated_info)
            save_inventory(inventory)
            print(f"Product with ID {product_id} updated successfully.")
            return
    print("Product ID not found.")


def search_product(name=None, category=None):
    inventory = load_inventory()
    results = []
    for product in inventory:
        if name and name.lower() in product['Name'].lower():
            results.append(product)
        elif category and category.lower() in product['Category'].lower():
            results.append(product)
    return results


def generate_reports():
    inventory = load_inventory()
    category_breakdown = {}
    total_value = 0
    for product in inventory:
        category = product['Category']
        value = product['Price'] * product['Quantity']
        total_value += value
        if category in category_breakdown:
            category_breakdown[category] += value
        else:
            category_breakdown[category] = value

    print("Category Breakdown:")
    for category, value in category_breakdown.items():
        print(f"Category: {category}, Total Value: {value}")
    print(f"Total Inventory Value: {total_value}")

# Initialize the inventory file
initialize_inventory()

# Adding initial 50 products and trying some edge cases

def generate_random_product(id):
    # Randomly generate a product name
    product_name = 'Product' + ''.join(random.choices(string.ascii_uppercase, k=5))

    # Randomly select a category (e.g., Category1, Category2, etc.)
    category = 'Category' + str(random.randint(1, 5))

    # Randomly generate a price between 5 and 500
    price = round(random.uniform(5.0, 500.0), 2)

    # Randomly generate a quantity between 1 and 100
    quantity = random.randint(1, 100)

    return {
        "ID": id,
        "Name": product_name,
        "Category": category,
        "Price": price,
        "Quantity": quantity
    }

def generate_and_add_products():
    # Generate and add 50 random products
    for i in range(1, 51):
        random_product = generate_random_product(i)
        add_product(random_product)

# Initialize the inventory file
initialize_inventory()

# Generate and add random products
generate_and_add_products()

# You can now also generate reports or perform other operations like deletion, updates, etc.
generate_reports()
# Testing edge cases
add_product({"ID": 1, "Name": "DuplicateProduct", "Category": "Category2", "Price": 15.0, "Quantity": 10})
delete_product(999)  # Non-existent product
update_product(999, {"Name": "UpdatedName"})  # Update non-existent product
search_product(name="NonExistentProduct")

Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Please use a unique ID.
Product ID already exists. Plea

[]

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


# Question 7 (20 points)

Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

In [ ]:
import random

class Flight:
    def __init__(self, flight_id, origin, destination, seats):
        self.flight_id = flight_id
        self.origin = origin
        self.destination = destination
        self.total_seats = seats
        self.available_seats = seats
        self.reservations = {}

    def reserve_seat(self, passenger_name):
        if self.available_seats > 0:
            if passenger_name in self.reservations:
                return f"Passenger {passenger_name} already has a reservation."
            seat_number = self.total_seats - self.available_seats + 1
            self.reservations[passenger_name] = seat_number
            self.available_seats -= 1
            return f"Seat {seat_number} reserved for {passenger_name} on flight {self.flight_id}."
        else:
            return "No available seats on this flight. Overbooking is not allowed."

    def release_seat(self, passenger_name):
        if passenger_name in self.reservations:
            self.available_seats += 1
            del self.reservations[passenger_name]
            return f"Reservation for {passenger_name} has been cancelled."
        else:
            return "No reservation found for the given passenger."

    def __str__(self):
        return f"Flight {self.flight_id} ({self.origin} -> {self.destination}) - Available Seats: {self.available_seats}/{self.total_seats}"


class FlightBookingSystem:
    def __init__(self):
        self.flights = {}
        self.initialize_flights()

    def initialize_flights(self):
        origins = ["New York", "Los Angeles", "Chicago", "Dallas", "Miami"]
        destinations = ["London", "Tokyo", "Paris", "Berlin", "Sydney"]
        for i in range(1, 51):
            origin = random.choice(origins)
            destination = random.choice(destinations)
            while origin == destination:
                destination = random.choice(destinations)
            seats = random.randint(50, 200)
            flight = Flight(f"F{i:03}", origin, destination, seats)
            self.flights[flight.flight_id] = flight

    def find_flight(self, origin, destination):
        return [flight for flight in self.flights.values() if flight.origin == origin and flight.destination == destination]

    def book_seat(self, flight_id, passenger_name):

        if flight_id in self.flights.keys():
            return self.flights[flight_id].reserve_seat(passenger_name)
        else:

            return "Invalid flight ID. Please enter a valid flight number."

    def cancel_reservation(self, flight_id, passenger_name):
        if flight_id in self.flights:
            return self.flights[flight_id].release_seat(passenger_name)
        else:
            return "Invalid flight ID. Please enter a valid flight number."

    def generate_report(self):
        report = "Flight Booking System Report:\n"
        for flight in self.flights.values():
            report += str(flight) + "\n"
        return report

# Example Usage
system = FlightBookingSystem()
print(system.generate_report())

# Valid Booking and Cancellation

print(system.cancel_reservation("F001", "Alice"))

# Invalid Flight ID
print(system.book_seat("dvb", "Bob"))  # Should return invalid flight ID error
print(system.cancel_reservation("sam", "Charlie"))  # Should return invalid flight ID error

# Overbooking Scenario
for _ in range(210):  # Attempt to overbook beyond capacity
    print(system.book_seat("F001", f"Passenger{_}"))

# Duplicate Booking Attempt
  # Should return duplicate reservation error

# Cancellation for Non-existent Passenger
print(system.cancel_reservation("F002", "Unknown"))  # Should return no reservation found


Flight Booking System Report:
Flight F001 (Chicago -> Berlin) - Available Seats: 190/190
Flight F002 (New York -> Sydney) - Available Seats: 129/129
Flight F003 (Los Angeles -> Sydney) - Available Seats: 87/87
Flight F004 (Miami -> London) - Available Seats: 190/190
Flight F005 (Los Angeles -> Paris) - Available Seats: 70/70
Flight F006 (Miami -> Berlin) - Available Seats: 58/58
Flight F007 (Miami -> Sydney) - Available Seats: 127/127
Flight F008 (Dallas -> Sydney) - Available Seats: 57/57
Flight F009 (Chicago -> Berlin) - Available Seats: 79/79
Flight F010 (Dallas -> Berlin) - Available Seats: 190/190
Flight F011 (Los Angeles -> Berlin) - Available Seats: 188/188
Flight F012 (Dallas -> Paris) - Available Seats: 91/91
Flight F013 (Dallas -> Sydney) - Available Seats: 152/152
Flight F014 (Miami -> Berlin) - Available Seats: 138/138
Flight F015 (Dallas -> Tokyo) - Available Seats: 70/70
Flight F016 (Chicago -> Tokyo) - Available Seats: 125/125
Flight F017 (Los Angeles -> Paris) - Availab

# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below